compiler/rustc_data_structures/src/sync/freeze.rs

```diff
@@ -0,0 +1,157 @@
+use crate::sync::{AtomicBool, Lock, LockGuard};
+#[cfg(parallel_compiler)]
+use crate::sync::{DynSend, DynSync};
+use std::{
+    cell::UnsafeCell,
+    intrinsics::likely,
+    ops::{Deref, DerefMut},
+    sync::atomic::Ordering,
+};
+
+/// A type which allows mutation using a lock until
+/// the value is frozen and can be accessed lock-free.
+#[derive(Default)]
+pub struct Freeze<T> {
```

Comment:
When I read the PR title I thought you were talking about the internal `std::marker::Freeze` trait.

Reply:
I renamed it to `FreezeLock` to avoid confusing it with that.